simulator/infra/macro.h

```diff
@@ -280,4 +280,74 @@ static inline uint128 gen_reverse( uint128 /* src1 */, size_t /* shamt */) {
     throw std::runtime_error( "Generalized reverse is not implemented for RV128");
     return 0;
 }
+
+uint64_t shuffle_stage( uint64_t src, uint64_t maskL, uint64_t maskR, size_t N)
+{
+    uint64_t x = src & ~( maskL | maskR);
+    x |= ( ( src << N) & maskL) | ( ( src >> N) & maskR);
+    return x;
+}
+
+// array with constants for shuffle
+std::array<uint64_t> shfl_consts[] = {
+    0x0000000000ff0000ull, 0x000000000000ff00ull,
+    0x000000000f000f00ull, 0x0000000000f000f0ull,
+    0x0000000030303030ull, 0x000000000c0c0c0cull,
+    0x0000000044444444ull, 0x0000000022222222ull,
+    0x0000ffff00000000ull, 0x00000000ffff0000ull,
+    0x00ff000000ff0000ull, 0x0000ff000000ff00ull,
+    0x0f000f000f000f00ull, 0x00f000f000f000f0ull,
+    0x3030303030303030ull, 0x0c0c0c0c0c0c0c0cull,
+    0x4444444444444444ull, 0x2222222222222222ull,
+};
+
+template<typename T> static
+auto bit_shuffle( T src1, T src2)
+{
+    auto pre_result = src1;
+    auto shamt = src2 & ( ( bytewidth<T> > bytewidth<uint32_t>) ? 31 : 15);
+    size_t offset = bytewidth<T> / bytewidth<uint64_t>;
+    size_t limit = 4 + offset;
+    for( size_t i = 0; i < limit ; ++i)
+    {
+        if ( shamt & ( 1 << ( limit - i - 1)))
+            pre_result = shuffle_stage( pre_result,
+                                            shfl_consts[2 * i + offset * 8],
+                                            shfl_consts[2 * i + offset * 8 + 1], 
+                                            1 << ( limit - i - 1));
+    }
+    return pre_result;
```

Comment:
Probably, having 3 overloads would make a simpler code. Currently it's quite hard to distinguish between 32 and 64 bit cases.